File: backend/app/schemas/translation.py
"""Pydantic schemas for the multilingual content platform (Phase 5.1).

Kept alongside the legacy AI-translation request/response DTOs — those
serve the ``/api/v1/translation`` free-text endpoint and are unchanged.
Everything below powers the per-entity translation platform.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import IdentifiedDto


# -- Legacy free-text translation DTOs (unchanged) ---------------------------


class TranslateRequest(BaseModel):
    text: str = Field(min_length=1, max_length=20000)
    sourceLanguage: str | None = None
    targetLanguage: str = Field(min_length=2, max_length=10)
    workspaceId: str | None = None


class TranslationDto(IdentifiedDto):
    sourceLanguage: str
    targetLanguage: str
    sourceText: str
    translatedText: str
    quality: float


# -- Enum literals (mirror app.constants.translation) ------------------------

TranslationStatus = Literal["draft", "translated", "reviewed", "published"]
JobStatus = Literal["pending", "processing", "completed", "failed", "cancelled"]
SupportedEntityType = Literal[
    "disaster", "public_resource", "campaign", "organization"
]


# -- Entity translation ------------------------------------------------------


class EntityTranslationBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    entityType: SupportedEntityType = Field(alias="entity_type")
    entityId: uuid.UUID = Field(alias="entity_id")
    locale: str = Field(min_length=2, max_length=20)
    fieldName: str = Field(alias="field_name", min_length=1, max_length=80)


class EntityTranslationCreate(EntityTranslationBase):
    translatedValue: str = Field(alias="translated_value", default="")
    status: TranslationStatus = "draft"
    sourceHash: str | None = Field(alias="source_hash", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EntityTranslationUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    translatedValue: str | None = Field(alias="translated_value", default=None)
    status: TranslationStatus | None = None
    sourceHash: str | None = Field(alias="source_hash", default=None)
    reviewedByUserId: uuid.UUID | None = Field(
        alias="reviewed_by_user_id", default=None
    )
    metadata: dict[str, Any] | None = None


class EntityTranslationDto(IdentifiedDto):
    entityType: str = Field(alias="entity_type")
    entityId: uuid.UUID = Field(alias="entity_id")
    locale: str
    fieldName: str = Field(alias="field_name")
    translatedValue: str = Field(alias="translated_value")
    status: TranslationStatus
    sourceHash: str | None = Field(alias="source_hash", default=None)
    translatedByUserId: uuid.UUID | None = Field(
        alias="translated_by_user_id", default=None
    )
    reviewedByUserId: uuid.UUID | None = Field(
        alias="reviewed_by_user_id", default=None
    )
    metadata: dict[str, Any] = Field(default_factory=dict)


class EntityTranslationListQuery(BaseModel):
    entityType: SupportedEntityType | None = None
    entityId: uuid.UUID | None = None
    locale: str | None = None
    status: TranslationStatus | None = None
    fieldName: str | None = None
    page: int = 1
    pageSize: int = 50


class EntityTranslationSearchQuery(BaseModel):
    query: str = Field(min_length=1, max_length=200)
    entityType: SupportedEntityType | None = None
    locale: str | None = None
    status: TranslationStatus | None = None
    page: int = 1
    pageSize: int = 50


# -- Translation job ---------------------------------------------------------


class TranslationJobCreate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    entityType: SupportedEntityType = Field(alias="entity_type")
    entityId: uuid.UUID = Field(alias="entity_id")
    sourceLocale: str = Field(alias="source_locale")
    targetLocale: str = Field(alias="target_locale")
    provider: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class TranslationJobDto(IdentifiedDto):
    entityType: str = Field(alias="entity_type")
    entityId: uuid.UUID = Field(alias="entity_id")
    sourceLocale: str = Field(alias="source_locale")
    targetLocale: str = Field(alias="target_locale")
    status: JobStatus
    provider: str | None = None
    requestedByUserId: uuid.UUID | None = Field(
        alias="requested_by_user_id", default=None
    )
    requestedAt: datetime | None = Field(alias="requested_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)


# -- Translation locale ------------------------------------------------------


class TranslationLocaleCreate(BaseModel):
    locale: str = Field(min_length=2, max_length=20)
    displayName: str = Field(alias="display_name", min_length=1, max_length=120)
    nativeName: str | None = Field(alias="native_name", default=None)
    rtl: bool = False
    enabled: bool = True
    defaultLocale: bool = Field(alias="default_locale", default=False)
    sortOrder: int = Field(alias="sort_order", default=0)

    model_config = ConfigDict(populate_by_name=True)


class TranslationLocaleUpdate(BaseModel):
    displayName: str | None = Field(alias="display_name", default=None)
    nativeName: str | None = Field(alias="native_name", default=None)
    rtl: bool | None = None
    enabled: bool | None = None
    defaultLocale: bool | None = Field(alias="default_locale", default=None)
    sortOrder: int | None = Field(alias="sort_order", default=None)

    model_config = ConfigDict(populate_by_name=True)


class TranslationLocaleDto(IdentifiedDto):
    locale: str
    displayName: str = Field(alias="display_name")
    nativeName: str | None = Field(alias="native_name", default=None)
    rtl: bool
    enabled: bool
    defaultLocale: bool = Field(alias="default_locale")
    sortOrder: int = Field(alias="sort_order")

File: src/types/translation.ts
// ============================================================================
// Translation platform types
//
// Two coexisting surfaces:
//   1. Legacy AI free-text translation (Translate / Batch) — kept for the
//      /translation workspace page.
//   2. Multilingual Content Platform (Phase 5) — per-entity translations,
//      jobs and locales. Consumed by /translations pages and services.
// ============================================================================

// ── Legacy free-text translation ────────────────────────────────────────────

export interface TranslationQualityScores {
  accuracy: number;
  readability: number;
  tone: number;
  consistency: number;
  grammar: number;
  accessibility: number;
}

export interface TranslationRequest {
  sourceLanguage: string;
  targetLanguage: string;
  content: string;

  // Required by backend AI translation endpoint
  workspaceId?: string;

  // Optional legacy options
  glossaryId?: string;
  preserveVariables?: boolean;
}

export interface TranslationResult {
  id: string;

  sourceLanguage: string;
  targetLanguage: string;

  sourceContent: string;
  translatedContent: string;

  scores: TranslationQualityScores;

  createdAt: string;
  createdBy: string;

  characters: number;
}

export interface BatchTranslationRequest {
  sourceLanguage: string;
  targetLanguages: string[];

  content: string;

  // Passed through to every translate() call
  workspaceId?: string;
}

export interface BatchTranslationResult {
  id: string;
  sourceLanguage: string;
  entries: TranslationResult[];
  createdAt: string;
}

export interface GlossaryTerm {
  id: string;
  source: string;
  translations: Record<string, string>;
  category?: string;
  notes?: string;
}

export interface Glossary {
  id: string;
  name: string;
  description?: string;
  termCount: number;
  updatedAt: string;
}
// ── Multilingual Content Platform ───────────────────────────────────────────

export const TRANSLATION_STATUSES = [
  "draft",
  "translated",
  "reviewed",
  "published",
] as const;
export type TranslationStatus = (typeof TRANSLATION_STATUSES)[number];

export const TRANSLATION_ENTITY_TYPES = [
  "disaster",
  "public_resource",
  "campaign",
  "organization",
] as const;
export type TranslationEntityType = (typeof TRANSLATION_ENTITY_TYPES)[number];

export const TRANSLATION_JOB_STATUSES = [
  "pending",
  "processing",
  "completed",
  "failed",
  "cancelled",
] as const;
export type TranslationJobStatus = (typeof TRANSLATION_JOB_STATUSES)[number];

export interface EntityTranslation {
  id: string;
  entityType: TranslationEntityType | string;
  entityId: string;
  locale: string;
  fieldName: string;
  translatedValue: string;
  status: TranslationStatus;
  sourceHash: string | null;
  translatedByUserId: string | null;
  reviewedByUserId: string | null;
  metadata: Record<string, unknown>;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface EntityTranslationInput {
  entityType: TranslationEntityType | string;
  entityId: string;
  locale: string;
  fieldName: string;
  translatedValue: string;
  status?: TranslationStatus;
  sourceHash?: string | null;
  metadata?: Record<string, unknown>;
}

export interface EntityTranslationUpdate {
  translatedValue?: string;
  status?: TranslationStatus;
  sourceHash?: string | null;
  metadata?: Record<string, unknown>;
}

export interface TranslationListQuery {
  page?: number;
  pageSize?: number;
  query?: string;
  entityType?: string;
  entityId?: string;
  locale?: string;
  status?: TranslationStatus;
  fieldName?: string;
  translatorId?: string;
  reviewerId?: string;
  sortBy?: string;
  sortDir?: "asc" | "desc";
}

export interface TranslationJob {
  id: string;
  entityType: string;
  entityId: string;
  sourceLocale: string;
  targetLocale: string;
  status: TranslationJobStatus;
  provider: string | null;
  requestedByUserId: string | null;
  requestedAt: string | null;
  completedAt: string | null;
  metadata: Record<string, unknown>;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface TranslationJobInput {
  entityType: TranslationEntityType | string;
  entityId: string;
  sourceLocale: string;
  targetLocale: string;
  provider?: string;
  metadata?: Record<string, unknown>;
}

export interface TranslationJobListQuery {
  page?: number;
  pageSize?: number;
  entityType?: string;
  entityId?: string;
  status?: TranslationJobStatus;
  targetLocale?: string;
  requestedByUserId?: string;
  sortBy?: string;
  sortDir?: "asc" | "desc";
}

export interface TranslationLocale {
  id: string;
  locale: string;
  displayName: string;
  nativeName: string | null;
  rtl: boolean;
  enabled: boolean;
  defaultLocale: boolean;
  sortOrder: number;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface TranslationLocaleInput {
  locale: string;
  displayName: string;
  nativeName?: string | null;
  rtl?: boolean;
  enabled?: boolean;
  defaultLocale?: boolean;
  sortOrder?: number;
}

export interface TranslationLocaleUpdate {
  displayName?: string;
  nativeName?: string | null;
  rtl?: boolean;
  enabled?: boolean;
  defaultLocale?: boolean;
  sortOrder?: number;
}
